Styles updates, Theming and other misc updates

aulneau opened the pull request:
Okay, I've gone through just about every style in this thing and updated / refactored them to be cleaner and use variables whenever possible.

I've also added a new step to our css, I've added postcss to process our generated css and do a few different things. It seemed like we weren't autoprefixing our css at all, so I've added `autoprefixer`. I've also opted to use `cssnano` with postcss to minify our css in that process, so we are no longer using the `node-sass compress` functionality. I've added a postcss plugin to combine duplicate selectors into one.

For IE11, I've added a couple postcss plugins to convert all of our custom css-props to their true values, so that if people need to support IE11 (eg: chart studio), you just include the ie11 stylesheet instead. This step can also be integrated into the users app, so that if they wanted to theme the editor first, and then run the postcss plugin themselves, they could. I've also added a plugin to remove `:root{}` from the stylesheet if you are generating styles for ie11. 

I've also added an empty state for if there is no trace data, the message could/should be modified:

![screen shot 2017-12-31 at 4 35 42 pm](https://user-images.githubusercontent.com/11803153/34464529-b10e9134-ee48-11e7-9a2c-acd9ea8e84fa.png)

I have added a delete component within the field component:

![screen shot 2017-12-31 at 4 36 27 pm](https://user-images.githubusercontent.com/11803153/34464530-c54d070c-ee48-11e7-9a08-94cf0d7d2833.png)

I've also added a Button component.

I still need to go through and update some icons to what we have in `plotly-icons`, but I wanted to get this PR open :)

Regarding theming: I've opted to use scss maps and then parse through them to generate our scss variables.

The various scss maps:

<img width="461" alt="screen shot 2018-01-02 at 10 53 06 am" src="https://user-images.githubusercontent.com/11803153/34491941-2b033db0-efab-11e7-8cc4-a09c95bfc35d.png">

are mapped through with scss functions:

<img width="670" alt="screen shot 2018-01-02 at 10 53 18 am" src="https://user-images.githubusercontent.com/11803153/34491956-33b4b902-efab-11e7-8ff3-79b3783d0f23.png">

which then are outputted as css variables:

![screen shot 2018-01-02 at 10 52 00 am](https://user-images.githubusercontent.com/11803153/34491970-44aeb44c-efab-11e7-8cb2-c87a0e3ad1dc.png)

which can be easily overridden by the end user in their app:

<img width="652" alt="screen shot 2018-01-02 at 10 54 29 am" src="https://user-images.githubusercontent.com/11803153/34491992-57389e7a-efab-11e7-83d2-b6380e378645.png">

which allows for a themable app:

![screen shot 2018-01-02 at 10 55 15 am](https://user-images.githubusercontent.com/11803153/34492022-77fa83d0-efab-11e7-8e21-4303d6c7c07c.png)


resolves: #184 #183 #76

bpostlethwaite commented:
Awesome work! I haven't looked in depth at this yet and maybe @VeraZab can do a review of the CSS? 
As part of this PR could you add language to the README that tells people how they would customize this library? That and the build process are what I will focus on (the style and build API for developers). 

Also AFAIK you are not restricted to NODE_ENV. You could, for example, use another env variable called PLOTLYJS_EDITOR_CSS_TARGET and then set that in the build command before running the javascript that does the building. 

NODE_ENV is usually used to distinguish between 'dev' and 'production' builds.

bpostlethwaite commented:
Also this PR seems to address more than one existing issue. I see "empty state for not traces". Could you update your description to include "resolves #x #y and #z" type declarations? Github will auto-close the issues and it will give me more context. Thanks!

VeraZab commented on package.json:
```diff
@@ -79,14 +79,20 @@
+    "es6-shim": "^0.35.3",
     "fast-isnumeric": "^1.1.1",
+    "mdi-react": "^2.1.19",
     "plotly-icons": "^1.0.1",
```
@aulneau you're going to add the additional mdi-react icons we need to plotly-icons?

VeraZab commented on src/components/containers/TraceAccordion.js:
```diff
@@ -27,9 +29,13 @@ export default class TraceAccordion extends Component {
+        className="js-add-trace-button"
+        variant="primary"
+        onClick={this.addTrace}
+        icon={<PlusIcon />}
```
or PlusIcon could be the default in the Button component?

VeraZab replied:
or maybe not, if we're including modals to the editor (which I'm not sure we will be), maybe we want more flexibility on that button

aulneau replied:
Maybe... I think it might be better to leave no default icon, because there are buttons that don't have icons.

VeraZab commented on src/components/sidebar/SidebarGroup.js:
```diff
@@ -1,7 +1,7 @@
-
+import {bem} from 'lib';
+import ChevronRightIcon from 'mdi-react/ChevronRightIcon';
 import SidebarItem from './SidebarItem';
```
it's AngleDownIcon in plotly-icons

aulneau replied:
Cool, I'll make sure to update all of the icons throughout with our `plotly-icons` and add any I use throughout :)

VeraZab commented on src/styles/_helpers.scss:
```diff
@@ -1,3 +1,33 @@
+    @return map-get($spacings,$name);
+  } @else {
+    @return var(--spacing-#{$name});
+  }
```
is this a sass thing: `var(--spacing-#{$name})` how does it work?

aulneau replied:
whoops, these are no longer used, I'll remove them. 

I've added a lot of scss maps and parsers, to out put css custom properties (native css variables) from our scss.

VeraZab commented on src/styles/_mixins.scss:
```diff
@@ -34,7 +34,7 @@
 
-@mixin scrollbar($width: 5px, $trackbg: $color-page-background, $thumb-color: $color-accent) {
+@mixin scrollbar($width: 5px, $trackbg: var(--scrollbar-track-background), $thumb-color: var(--scrollbar-thumb-color)) {
   -webkit-overflow-scrolling: touch;
```
what's the difference here between `$color-accent` and `var(--scrollbar-thumb-color)`

aulneau replied:
So a lot of these variables are kind of passed around so that you can theme the application, the reason this is set this way is so that the end user can change the accent color, and the scollbar default thumb color would be the accent color, but if they wanted a different color for the scrollbar, they would change the value of `--scrollbar-thumb-color`.

aulneau commented:
Yeah I know :/ I felt so bad making like 6 more scripts in there... There has to be a way to abstract and make it more efficient.

bpostlethwaite commented on src/components/containers/TraceAccordion.js:
```diff
@@ -40,8 +46,23 @@ export default class TraceAccordion extends Component {
+          message={
+            <p>
+              Looks like there aren't any traces defined yet. Go to the 'Create'
+              tab to define some traces.
```
this needs to be wrapped in i18n `_()`. Ping me or @VeraZab with any questions

bpostlethwaite commented on src/components/containers/TraceAccordion.js:
```diff
@@ -27,9 +29,13 @@ export default class TraceAccordion extends Component {
+        variant="primary"
+        onClick={this.addTrace}
+        icon={<PlusIcon />}
+        label="Trace"
```
this needs to be wrapped in i18n

bpostlethwaite commented on src/components/containers/AnnotationAccordion.js:
```diff
@@ -30,14 +31,18 @@ export default class AnnotationAccordion extends Component {
+        className="js-add-annotation-button"
+        onClick={this.addAnnotation}
+        icon={<PlusIcon />}
+        label="Annotation"
```
i18n

bpostlethwaite commented on src/components/Button.js:
```diff
@@ -0,0 +1,44 @@
+import PropTypes from 'prop-types';
```
Button should go in `components/widgets`?

bpostlethwaite commented on src/styles/components/_button.scss:
```diff
@@ -0,0 +1,73 @@
+// button default styles
```
this should be within a folder in components/? Seems like a widget to me.

aulneau commented:
Okay @bpostlethwaite and @VeraZab, I've updated the readme, added a theming document, and I believe I've fixed everything you two have requested. If you can review my latest changes, I'd say this is good to go! :) (barring any new changes requested)

VeraZab commented:
yey 💃

bpostlethwaite commented on package.json:
```diff
@@ -13,27 +13,29 @@
-      "mkdirp dist && browserify src/PlotlyEditor.js -o ./dist/PlotlyEditor.js -t [ babelify --presets [ es2015 react ] ] -t browserify-global-shim --standalone createPlotlyComponent && uglifyjs ./dist/PlotlyEditor.js --compress --mangle --output ./dist/PlotlyEditor.min.js --source-map filename=dist/PlotlyEditor.min.js.map && make:dist:css",
-    "make:dist:css":
-      "mkdirp dist && node-sass --output-style compressed src/styles/main.scss > dist/react-plotly.js-editor.css",
+    "make:lib": "mkdirp lib && babel src --out-dir lib --ignore=__tests__/* --source-maps && npm run make:lib:css",
```
@VeraZab I think you and @aulneau have squabbling .json formatting. I don't think we should use prettier on anything outside `/src` or at least not on anything other than *.js files.

aulneau replied:
Yeah, sometimes my muscle memory causes me to autoformat anything after I hit save. But it's strange, my current package.json is formatted correctly:

<img width="1211" alt="screen shot 2018-01-03 at 12 03 03 pm" src="https://user-images.githubusercontent.com/11803153/34533057-16fcc9e8-f07e-11e7-86ed-308b82c1ba36.png">

bpostlethwaite commented:
thanks for the heroic level of effort here! Can't wait to try out some new themes! 

:dancer:

VeraZab commented:
new release after this merge! : )

aulneau commented:
@VeraZab okay I fixed the `prepublishOnly` task. should be good to go! (also I don't have permission to merge this PR)
